processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java

```diff
@@ -0,0 +1,520 @@
+        return bufferComparator.compare(tableBuffer, tableBuffer, o1 + HASH_SIZE, o2 + HASH_SIZE);
+      }
+    };
+  }
+
+
+  private void validateBufferCapacity(
+      int limit,
+      float maxLoadFactor,
+      ByteBuffer buffer,
+      int bucketSize
+  )
+  {
+    int numBucketsNeeded = (int) Math.ceil((limit + 1) / maxLoadFactor);
+    int targetTableArenaSize = numBucketsNeeded * bucketSize * 2;
+    int heapSize = (limit + 1) * (Ints.BYTES);
+    int requiredSize = targetTableArenaSize + heapSize;
+
+    if (buffer.capacity() < requiredSize) {
+      throw new IAE(
+          "Buffer capacity [%d] is too small for limit[%d] with load factor[%f], minimum bytes needed: [%d]",
+          buffer.capacity(),
+          limit,
+          maxLoadFactor,
+          requiredSize
+      );
+    }
+  }
+
+  private class AlternatingByteBufferHashTable extends ByteBufferHashTable
+  {
+    // The base buffer is split into two alternating halves, with one sub-buffer in use at a given time.
+    // When the current sub-buffer fills, the used bits of the other sub-buffer are cleared, entries up to the limit
+    // are copied from the current full sub-buffer to the new buffer, and the active buffer (referenced by tableBuffer)
+    // is swapped to the new buffer.
+    private ByteBuffer[] subHashTableBuffers;
+
+    public AlternatingByteBufferHashTable(
+        float maxLoadFactor,
+        int initialBuckets,
+        int bucketSizeWithHash,
+        ByteBuffer totalHashTableBuffer,
+        int keySize,
+        int maxSizeForTesting
+    )
+    {
+      super(
+          maxLoadFactor,
+          initialBuckets,
+          bucketSizeWithHash,
+          totalHashTableBuffer,
+          keySize,
+          maxSizeForTesting,
+          null
+      );
+
+      this.growthCount = 0;
+
+      int subHashTableSize = tableArenaSize / 2;
+      maxBuckets = subHashTableSize / bucketSizeWithHash;
+      regrowthThreshold = maxSizeForBuckets(maxBuckets);
+
+      // split the hashtable into 2 sub tables that we rotate between
+      ByteBuffer subHashTable1Buffer = totalHashTableBuffer.duplicate();
+      subHashTable1Buffer.position(0);
+      subHashTable1Buffer.limit(subHashTableSize);
+      subHashTable1Buffer = subHashTable1Buffer.slice();
+
+      ByteBuffer subHashTable2Buffer = totalHashTableBuffer.duplicate();
+      subHashTable2Buffer.position(subHashTableSize);
+      subHashTable2Buffer.limit(tableArenaSize);
+      subHashTable2Buffer = subHashTable2Buffer.slice();
+
+      subHashTableBuffers = new ByteBuffer[] {subHashTable1Buffer, subHashTable2Buffer};
+    }
+
+    @Override
+    public void reset()
+    {
+      size = 0;
+      growthCount = 0;
+      // clear the used bits of both buffers
+      for (int i = 0; i < maxBuckets; i++) {
+        subHashTableBuffers[0].put(i * bucketSizeWithHash, (byte) 0);
+        subHashTableBuffers[1].put(i * bucketSizeWithHash, (byte) 0);
```

Comment:
this one is not needed and will automatically be zeroed out in adjustTableWhenFull(..) when needed.